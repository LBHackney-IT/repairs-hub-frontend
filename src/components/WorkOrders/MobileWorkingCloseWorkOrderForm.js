import PropTypes from 'prop-types'
import { useForm } from 'react-hook-form'
import BackButton from '../Layout/BackButton'
import TextArea from '../Form/TextArea'
import { PrimarySubmitButton } from '../Form'
import { useState } from 'react'
import FollowOnRequestMaterialsForm from './CloseWorkOrderFormComponents/FollowOnRequestMaterialsForm'
import FollowOnRequestTypeOfWorkForm from './CloseWorkOrderFormComponents/FollowOnRequestTypeOfWorkForm'
import CloseWorkOrderFormReasonForClosing from './CloseWorkOrderFormComponents/CloseWorkOrderFormReasonForClosing'
import validateFileUpload from '../WorkOrder/Photos/hooks/validateFileUpload'
import ControlledFileInput from '../WorkOrder/Photos/ControlledFileInput'
import CheckboxSmall from '../Form/CheckboxSmall'

const PAGES = {
  WORK_ORDER_STATUS: '1',
  FOLLOW_ON_DETAILS: '2',
}

const FIELD_NAMES_ON_FIRST_PAGE = [
  'reason',
  'followOnStatus',
  'fileUpload',
  'description',
]

const MobileWorkingCloseWorkOrderForm = ({ onSubmit, isLoading }) => {
  const {
    handleSubmit,
    register,
    errors,
    setError,
    clearErrors,
    watch,
    getValues,
    trigger,
  } = useForm({
    shouldUnregister: false,
  })

  const selectedFurtherWorkRequired =
    watch('followOnStatus') === 'furtherWorkRequired'

  const [currentPage, setCurrentPage] = useState(PAGES.WORK_ORDER_STATUS)

  const [closeWithoutPhotos, setCloseWithoutPhotos] = useState(false)
  // So we dont show the error immediately
  const [photosTouched, setPhotosTouched] = useState(false)

  const viewFollowOnDetailsPage = (e) => {
    e.preventDefault()

    try {
      trigger(FIELD_NAMES_ON_FIRST_PAGE)
    } catch (e) {
      console.error('Trigger failed', e)
    }

    if (Object.keys(errors).length > 0) return

    // validate file uploaded
    if (files.length === 0 && !closeWithoutPhotos) {
      // user must confirm submit without photos
      return
    }

    setCurrentPage(PAGES.FOLLOW_ON_DETAILS)
  }

  const viewWorkOrderStatusPage = () => {
    setCurrentPage(PAGES.WORK_ORDER_STATUS)
  }

  const [files, setFiles] = useState([])

  return (
    <div className="mobile-working-close-work-order-form">
      <BackButton
        // if on second page, override back button
        onClick={
          currentPage === PAGES.FOLLOW_ON_DETAILS
            ? viewWorkOrderStatusPage
            : null
        }
      />

      <form
        role="form"
        onSubmit={handleSubmit((data) => {
          if (files.length === 0 && !closeWithoutPhotos) {
            // user must confirm submit without photos
            return
          }

          onSubmit(data, files)
        })}
      >
        <div
          style={{
            display: currentPage === PAGES.WORK_ORDER_STATUS ? 'block' : 'none',
          }}
        >
          <h1 className="lbh-heading-h2">Close work order</h1>

          <CloseWorkOrderFormReasonForClosing
            register={register}
            errors={errors}
            watch={watch}
          />

          <div className="govuk-form-group lbh-form-group">
            <ControlledFileInput
              files={files}
              setFiles={setFiles}
              validationError={errors?.fileUpload?.message}
              isLoading={isLoading}
              register={register('fileUpload', {
                validate: () => {
                  setPhotosTouched(true)

                  const validation = validateFileUpload(files)

                  if (validation === null) return true
                  return validation
                },
              })}
            />

            {files.length > 0 && (
              <TextArea
                name="description"
                label="Photo description"
                showAsOptional
                register={register}
              />
            )}

            {files.length === 0 && photosTouched && (
              <div className="lbh-page-announcement lbh-page-announcement--warning">
                <h3 className="lbh-page-announcement__title">
                  No photos were selected
                </h3>
                <div className="lbh-page-announcement__content">
                  <div>
                    {' '}
                    Adding photos will help improve the identification of
                    follow-ons required and reduce errors.
                  </div>

                  <div style={{ marginTop: '30px !important' }}>
                    <CheckboxSmall
                      className="govuk-!-margin-0"
                      labelClassName="lbh-body-xs govuk-!-margin-0 govuk-!-margin-bottom-5"
                      name={'closeWorkOrderWithoutPhotos'}
                      label={'I want to close the work order without photos'}
                      onChange={() => {
                        setCloseWithoutPhotos(() => !closeWithoutPhotos)
                      }}
                      checked={closeWithoutPhotos}
                    />
                  </div>
                </div>
              </div>
            )}
          </div>

          <TextArea
            name="notes"
            label="Final report"
            register={register}
            error={errors && errors.notes}
          />

          {selectedFurtherWorkRequired ? (
            <PrimarySubmitButton
              label="Add details"
              type="button"
              onClick={viewFollowOnDetailsPage}
            />
          ) : (
            <PrimarySubmitButton label={`Close work order`} />
          )}
        </div>

        <div
          style={{
            display: currentPage === PAGES.FOLLOW_ON_DETAILS ? 'block' : 'none',
          }}
        >
          <h1 className="lbh-heading-h2">Details of further work required</h1>

          <FollowOnRequestTypeOfWorkForm
            errors={errors}
            register={register}
            getValues={getValues}
            setError={setError}
            clearErrors={clearErrors}
            watch={watch}
          />

          <FollowOnRequestMaterialsForm
            register={register}
            getValues={getValues}
            errors={errors}
          />

          <TextArea
            name="additionalNotes"
            label="Additional notes"
            register={register}
            error={errors && errors.additionalNotes}
          />

          <PrimarySubmitButton label="Close work order" />
        </div>
      </form>
    </div>
  )
}

MobileWorkingCloseWorkOrderForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
}

export default MobileWorkingCloseWorkOrderForm
