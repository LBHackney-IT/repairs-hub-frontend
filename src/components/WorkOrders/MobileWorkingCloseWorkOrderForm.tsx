import PropTypes from 'prop-types'
import { useForm } from 'react-hook-form'
import BackButton from '../Layout/BackButton'
import TextArea from '../Form/TextArea'
import { PrimarySubmitButton } from '../Form'
import { useState } from 'react'
import FollowOnRequestTypeOfWorkForm from './CloseWorkOrderFormComponents/FollowOnRequestTypeOfWorkForm'
import CloseWorkOrderFormReasonForClosing from './CloseWorkOrderFormComponents/CloseWorkOrderFormReasonForClosing'
import validateFileUpload from '../WorkOrder/Photos/hooks/validateFileUpload'
import ControlledFileInput from '../WorkOrder/Photos/ControlledFileInput'
import FollowOnRequestMaterialsSupervisorCalledForm from './CloseWorkOrderFormComponents/FollowOnRequestMaterialsSupervisorCalledForm'
import FollowOnRequestMaterialsForm from './CloseWorkOrderFormComponents/FollowOnRequestMaterialsForm'
import { SimpleFeatureToggleResponse } from '../../pages/api/simple-feature-toggle'

const PAGES = {
  WORK_ORDER_STATUS: '1',
  FOLLOW_ON_DETAILS: '2',
}

const FIELD_NAMES_ON_FIRST_PAGE = new Set<string>([
  'reason',
  'followOnStatus',
  'workOrderFileUpload',
])

interface MobileWorkingCloseWorkOrderFormProps {
  onSubmit: (
    data: Record<string, any>,
    workOrderFiles: File[],
    followOnFiles: File[]
  ) => void
  isLoading: boolean
}

const MobileWorkingCloseWorkOrderForm = ({
  onSubmit,
  isLoading,
}: MobileWorkingCloseWorkOrderFormProps) => {
  const {
    handleSubmit,
    register,
    errors,
    setError,
    clearErrors,
    watch,
    getValues,
    trigger,
  } = useForm({
    shouldUnregister: false,
  })

  const selectedFurtherWorkRequired =
    watch('followOnStatus') === 'furtherWorkRequired'

  const [currentPage, setCurrentPage] = useState(PAGES.WORK_ORDER_STATUS)

  const viewFollowOnDetailsPage = () => {
    trigger([...FIELD_NAMES_ON_FIRST_PAGE])

    if (
      Object.keys(errors).filter((key) => FIELD_NAMES_ON_FIRST_PAGE.has(key))
        .length > 0
    ) {
      console.info({ errors })
      return
    }
    setCurrentPage(PAGES.FOLLOW_ON_DETAILS)
  }

  const viewWorkOrderStatusPage = () => {
    setCurrentPage(PAGES.WORK_ORDER_STATUS)
  }

  const [workOrderFiles, setWorkOrderFiles] = useState([])
  const [followOnFiles, setFollowOnFiles] = useState([])

  return (
    <div className="mobile-working-close-work-order-form">
      <BackButton
        // if on second page, override back button
        onClick={
          currentPage === PAGES.FOLLOW_ON_DETAILS
            ? viewWorkOrderStatusPage
            : null
        }
      />

      <form
        role="form"
        onSubmit={handleSubmit((data) => {
          onSubmit(data, workOrderFiles, followOnFiles)
        })}
      >
        <div
          style={{
            display: currentPage === PAGES.WORK_ORDER_STATUS ? 'block' : 'none',
          }}
        >
          <h1 className="lbh-heading-h2">Close work order</h1>

          <CloseWorkOrderFormReasonForClosing
            register={register}
            errors={errors}
            watch={watch}
            canRaiseAFollowOn={true}
          />

          <div className="govuk-form-group lbh-form-group">
            <ControlledFileInput
              label="Work order photos"
              hint="Add photos showing the repair you completed (up to 10 photos)"
              files={workOrderFiles}
              setFiles={setWorkOrderFiles}
              validationError={errors?.workOrderFileUpload?.message}
              isLoading={isLoading}
              register={register('workOrderFileUpload', {
                validate: () => {
                  const validation = validateFileUpload(workOrderFiles)

                  if (validation === null) return true
                  return validation
                },
              })}
              testId="WorkOrderPhotoUpload"
            />

            {workOrderFiles.length > 0 && (
              <TextArea
                name="workOrderPhotoDescription"
                label="Photo description"
                register={register}
              />
            )}
          </div>

          <TextArea
            name="notes"
            label="Final report"
            register={register}
            error={errors && errors.notes}
          />

          {selectedFurtherWorkRequired ? (
            <PrimarySubmitButton
              label="Add details"
              type="button"
              onClick={viewFollowOnDetailsPage}
            />
          ) : (
            <PrimarySubmitButton
              id="submit-work-order-close"
              label="Close work order"
            />
          )}
        </div>

        {currentPage === PAGES.FOLLOW_ON_DETAILS && (
          <div>
            <h1 className="lbh-heading-h2">Details of further work required</h1>

            <FollowOnRequestMaterialsSupervisorCalledForm
              register={register}
              errors={errors}
            />

            <FollowOnRequestTypeOfWorkForm
              errors={errors}
              register={register}
              getValues={getValues}
              setError={setError}
              clearErrors={clearErrors}
              watch={watch}
            />
            <FollowOnRequestMaterialsForm
              register={register}
              getValues={getValues}
              errors={errors}
            />

            <TextArea
              name="additionalNotes"
              label="Additional notes"
              register={register}
              error={errors && errors.additionalNotes}
            />

            <div className="govuk-form-group lbh-form-group">
              <ControlledFileInput
                label="Follow on photos"
                hint="Add photos showing the follow on work needed (up to 10 photos)"
                files={followOnFiles}
                setFiles={setFollowOnFiles}
                validationError={errors?.followOnFileUpload?.message}
                isLoading={isLoading}
                register={register('followOnFileUpload', {
                  validate: () => {
                    const validation = validateFileUpload(followOnFiles)

                    if (validation === null) return true
                    return validation
                  },
                })}
                testId="FollowOnPhotoUpload"
              />

              {followOnFiles.length > 0 && (
                <TextArea
                  name="followOnPhotoDescription"
                  label="Photo description"
                  register={register}
                />
              )}
            </div>

            <PrimarySubmitButton
              id="submit-work-order-close"
              label="Close work order"
            />
          </div>
        )}
      </form>
    </div>
  )
}

MobileWorkingCloseWorkOrderForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
}

export default MobileWorkingCloseWorkOrderForm
