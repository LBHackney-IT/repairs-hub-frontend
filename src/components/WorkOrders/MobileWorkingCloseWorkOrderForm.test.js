import { render } from '@testing-library/react'
import MobileWorkingCloseWorkOrderForm from './MobileWorkingCloseWorkOrderForm'

describe.skip('MobileWorkingCloseWorkOrderForm component', () => {
  it('should render properly', () => {
    const { asFragment } = render(
      <MobileWorkingCloseWorkOrderForm onSubmit={jest.fn()} />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
