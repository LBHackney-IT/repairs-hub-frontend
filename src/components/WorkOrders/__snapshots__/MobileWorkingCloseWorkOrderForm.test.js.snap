// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileWorkingCloseWorkOrderForm component should render properly 1`] = `
<DocumentFragment>
  <div>
    <a
      class="govuk-back-link lbh-back-link govuk-!-display-none-print"
      role="button"
    >
      Back
    </a>
    <form
      role="form"
    >
      <div
        style="display: block;"
      >
        <h1
          class="lbh-heading-h2"
        >
          Close work order form
        </h1>
        <div
          class="govuk-form-group lbh-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="reason"
          >
            Select reason for closing 
          </label>
          <div
            class="govuk-radios lbh-radios"
          >
            <div
              class="govuk-radios__item"
            >
              <input
                class="govuk-radios__input"
                data-testid="reason"
                id="reason_Work Order Completed"
                name="reason"
                type="radio"
                value="Work Order Completed"
              />
              <label
                class="govuk-label lbh-label govuk-radios__label"
                for="reason_Work Order Completed"
              >
                Visit completed
              </label>
            </div>
            <div
              class="govuk-radios__conditional"
              id="conditional-contact"
            />
            <div
              class="govuk-radios__item"
            >
              <input
                class="govuk-radios__input"
                data-testid="reason"
                id="reason_No Access"
                name="reason"
                type="radio"
                value="No Access"
              />
              <label
                class="govuk-label lbh-label govuk-radios__label"
                for="reason_No Access"
              >
                No access
              </label>
            </div>
          </div>
        </div>
        <div
          class="govuk-form-group lbh-form-group"
          id="notes-form-group"
        >
          <label
            class="govuk-label lbh-label"
            for="notes"
          >
            Final report 
          </label>
          <textarea
            aria-describedby="notes-hint notes-error"
            class="govuk-textarea lbh-textarea"
            data-testid="notes"
            id="notes"
            name="notes"
            rows="4"
          />
        </div>
        <div
          class="govuk-form-group lbh-form-group"
        >
          <button
            class="govuk-button lbh-button"
            data-module="govuk-button"
            type="submit"
          >
            Close work order
          </button>
        </div>
      </div>
      <div
        style="display: none;"
      >
        <h1
          class="lbh-heading-h2"
        >
          Details of further work required
        </h1>
        <fieldset
          class="govuk-fieldset govuk-!-margin-bottom-2 govuk-!-padding-2 lbh-fieldset"
        >
          <div
            class="lbh-form-group"
          >
            <label
              class="govuk-label govuk-label--m"
              for=""
            >
              Type of work required
            </label>
            <div
              class="govuk-checkboxes__item govuk-!-margin-0 govuk-!-margin-bottom-5"
            >
              <input
                class="govuk-checkboxes__input"
                data-testid="isSameTrade"
                id="isSameTrade"
                name="isSameTrade"
                type="checkbox"
              />
              <label
                class="govuk-label govuk-checkboxes__label lbh-body-xs govuk-!-margin-0"
                for="isSameTrade"
              >
                Same trade
              </label>
            </div>
            <div
              class="govuk-checkboxes__item govuk-!-margin-0 govuk-!-margin-bottom-5"
            >
              <input
                class="govuk-checkboxes__input"
                data-testid="isDifferentTrades"
                id="isDifferentTrades"
                name="isDifferentTrades"
                type="checkbox"
              />
              <label
                class="govuk-label govuk-checkboxes__label lbh-body-xs govuk-!-margin-0"
                for="isDifferentTrades"
              >
                Different trade(s)
              </label>
            </div>
            <div
              class="govuk-checkboxes__item govuk-!-margin-0 govuk-!-margin-bottom-5"
            >
              <input
                class="govuk-checkboxes__input"
                data-testid="isMultipleOperatives"
                id="isMultipleOperatives"
                name="isMultipleOperatives"
                type="checkbox"
              />
              <label
                class="govuk-label govuk-checkboxes__label lbh-body-xs govuk-!-margin-0"
                for="isMultipleOperatives"
              >
                Multiple operatives
              </label>
            </div>
          </div>
        </fieldset>
        <div
          class="govuk-form-group lbh-form-group"
          id="followOnTypeDescription-form-group"
        >
          <label
            class="govuk-label lbh-label"
            for="followOnTypeDescription"
          >
            Describe work required 
          </label>
          <textarea
            aria-describedby="followOnTypeDescription-hint followOnTypeDescription-error"
            class="govuk-textarea lbh-textarea"
            data-testid="followOnTypeDescription"
            id="followOnTypeDescription"
            name="followOnTypeDescription"
            rows="4"
          />
        </div>
        <fieldset>
          <label
            class="govuk-label govuk-label--m"
          >
            Materials
          </label>
          <div
            class="govuk-checkboxes__item govuk-!-margin-0"
          >
            <input
              class="govuk-checkboxes__input"
              data-testid="stockItemsRequired"
              id="stockItemsRequired"
              name="stockItemsRequired"
              type="checkbox"
            />
            <label
              class="govuk-label govuk-checkboxes__label lbh-body-xs govuk-!-margin-0 govuk-!-margin-bottom-5"
              for="stockItemsRequired"
            >
              Stock items required
            </label>
          </div>
          <div
            class="govuk-checkboxes__item govuk-!-margin-0"
          >
            <input
              class="govuk-checkboxes__input"
              data-testid="nonStockItemsRequired"
              id="nonStockItemsRequired"
              name="nonStockItemsRequired"
              type="checkbox"
            />
            <label
              class="govuk-label govuk-checkboxes__label lbh-body-xs govuk-!-margin-0 govuk-!-margin-bottom-5"
              for="nonStockItemsRequired"
            >
              Non stock items required
            </label>
          </div>
        </fieldset>
        <div
          class="govuk-form-group lbh-form-group"
          id="materialNotes-form-group"
        >
          <label
            class="govuk-label lbh-label"
            for="materialNotes"
          >
            Materials required 
          </label>
          <textarea
            aria-describedby="materialNotes-hint materialNotes-error"
            class="govuk-textarea lbh-textarea"
            data-testid="materialNotes"
            id="materialNotes"
            name="materialNotes"
            rows="4"
          />
        </div>
        <div
          class="govuk-form-group lbh-form-group"
          id="additionalNotes-form-group"
        >
          <label
            class="govuk-label lbh-label"
            for="additionalNotes"
          >
            Additional notes 
          </label>
          <textarea
            aria-describedby="additionalNotes-hint additionalNotes-error"
            class="govuk-textarea lbh-textarea"
            data-testid="additionalNotes"
            id="additionalNotes"
            name="additionalNotes"
            rows="4"
          />
        </div>
        <div
          class="govuk-form-group lbh-form-group"
        >
          <button
            class="govuk-button lbh-button"
            data-module="govuk-button"
            type="submit"
          >
            Close work order
          </button>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
